### PR TITLE
Avoid putting secret in the exported JSON object

### PR DESCRIPTION
**Pull request by benmoriceau**
## What

The OSS deployment allows to do an export and import of the workspace configuration. If the users is not using a secret manager in their deployment, this will return API key and password in plain text. This is a serious security issue, especially since it has been presented by @timroes that some instance are exposed on the public internet. In order to avoid returning password or other values that should be in a secret we need to sanitize the export. This is what this PR is doing.

## Important files

- `DatabaseConfigPersistence.java`: Changes to call the sanitizer
- `DatabaseConfigPersistenceTest.java`: Test that the masking of the password is properly called.
- `FeatureFlag.java`: New feature flag for applying the sanitization or not
- `.env`: feature flag default value
- The rest of the file is the propagation of the `FeatureFlag` and `JsonSecretsProcessor` objects, in test and non test files.

### CODE REVIEW

File: airbyte-commons/src/main/java/io/airbyte/commons/features/EnvVariableFeatureFlags.java [hunk 1]
```diff
@@ -23,4 +23,9 @@ public boolean autoDisablesFailingConnections() {
     return Boolean.parseBoolean(System.getenv("AUTO_DISABLE_FAILING_CONNECTIONS"));
   }
 
+  @Override
```
Comment (cgardens) [inline airbyte-commons/src/main/java/io/airbyte/commons/features/EnvVariableFeatureFlags.java:26]: why do we have a separate class for `EnvVariableFeatureFlags` from `EnvConfigs`? Doesn't have to change now, but not clear to me why they are separate.

File: airbyte-commons/src/main/java/io/airbyte/commons/features/FeatureFlags.java [hunk 2]
```diff
@@ -14,4 +14,6 @@
 
   boolean autoDisablesFailingConnections();
 
+  boolean hideSecretsInExport();
```
Comment (cgardens) [inline airbyte-commons/src/main/java/io/airbyte/commons/features/FeatureFlags.java:17]: can we invert it? `exposeSecretsInExport` and make sure it defaults to `false`

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java [hunk 3]
```diff
@@ -1575,17 +1583,47 @@ public void replaceAllConfigs(final Map<AirbyteConfig, Stream<?>> configs, final
       result.put(ConfigSchema.STANDARD_SOURCE_DEFINITION.name(),
           standardSourceDefinitionWithMetadata
               .stream()
-              .map(ConfigWithMetadata::getConfig)
-              .map(Jsons::jsonNode));
+              .map(configWitMetadata -> {
```
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1586]: just to confirm, this doesn't mess up auto migration, right? i don't remember all the places `dumpConfigs` gets called.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java [hunk 4]
```diff
@@ -1575,17 +1583,47 @@ public void replaceAllConfigs(final Map<AirbyteConfig, Stream<?>> configs, final
...
+                if (!featureFlags.hideSecretsInExport()) {
+                  return Jsons.jsonNode(configWitMetadata.getConfig());
+                }
+
+                try {
+                  final StandardSourceDefinition standardSourceDefinition = getConfig(
+                      ConfigSchema.STANDARD_SOURCE_DEFINITION,
+                      configWitMetadata.getConfigId(),
+                      StandardSourceDefinition.class);
```
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1595]: why are you doing `getConfig(...)` here? isn't this going to result in n queries? if you just call `configWithMetdata.getConfig()` doesn't that get what you need with it already in memory?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java [hunk 5]
```diff
@@ -1575,17 +1583,47 @@ public void replaceAllConfigs(final Map<AirbyteConfig, Stream<?>> configs, final
...
+                  return Jsons.jsonNode(configWitMetadata.getConfig());
+                }
+
+                try {
+                  final StandardDestinationDefinition standardDestinationDefinition = getConfig(
+                      ConfigSchema.STANDARD_DESTINATION_DEFINITION,
+                      configWitMetadata.getConfigId(),
+                      StandardDestinationDefinition.class);
+                  final JsonNode connectionSpecs = standardDestinationDefinition.getSpec().getConnectionSpecification();
+                  final JsonNode sanitizedConfig = jsonSecretsProcessor.maskSecrets(Jsons.jsonNode(configWitMetadata.getConfig()), connectionSpecs);
```
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1621]: is this right? maybe i'm confused but this looks like it is sanitizing the destination definition not the destination itself. am i missing something?

**Review by cgardens — CHANGES_REQUESTED** [review]
approach makes sense. a few questions on the implementation.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java (hunk 4 above)
Comment (benmoriceau) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1595]: That's true, I think that I will need that when processing the `SOURCE_CONNECTION` and `DESTINATION_CONNECTION`. I will need to use the getConfig in order to be able to get the config schema which is used to mask the secrets.

File: airbyte-commons/src/main/java/io/airbyte/commons/features/EnvVariableFeatureFlags.java (hunk 1 above)
Comment (benmoriceau) [reply airbyte-commons/src/main/java/io/airbyte/commons/features/EnvVariableFeatureFlags.java:26]: It was to potentially support a change to a DB.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java (hunk 3 above)
Comment (benmoriceau) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1586]: I am not sure about what is the auto migration. From the usage I found with the find usage of IntelliJ, it looked ok. I will add reviewer to double check that.

Comment (cgardens) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1586]: did a little hunting and i agree, it should be safe.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java (hunk 5 above)
Comment (benmoriceau) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1621]: Same comment than the Source part. I have changed it.

**Comment by benmoriceau** [comment]
@cgardens I moved it back to draft, The test I am writing for this change is not passing and I still don't understand why.

**Comment by benmoriceau** [comment]
@cgardens I think it is ready to be reviewed. I have checked it in docker and the password was sanitize in the export of the config.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java [hunk 6]
```diff
@@ -1592,16 +1600,49 @@ public void replaceAllConfigs(final Map<AirbyteConfig, Stream<?>> configs, final
...
+              return Jsons.jsonNode(configWithMetadata.getConfig());
+            }
+
+            try {
+              final UUID destinationDefinition = configWithMetadata.getConfig().getDestinationDefinitionId();
+              final StandardDestinationDefinition standardDestinationDefinition = getConfig(
+                  ConfigSchema.STANDARD_DESTINATION_DEFINITION,
+                  destinationDefinition.toString(),
+                  StandardDestinationDefinition.class);
+              final JsonNode connectionSpecs = standardDestinationDefinition.getSpec().getConnectionSpecification();
```
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1637]: this can be singular. `connectionSpec`

**Review by cgardens — APPROVED** [review]

File: airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/DatabaseConfigPersistenceTest.java [hunk 7]
```diff
@@ -170,6 +178,27 @@ public void testDumpConfigs() throws Exception {
...
+    doReturn(new StandardDestinationDefinition()
+        .withSpec(mockedConnectorSpec)).when(configPersistence).getConfig(eq(ConfigSchema.STANDARD_DESTINATION_DEFINITION), any(), any());
+    doReturn(new StandardSourceDefinition()
+        .withSpec(mockedConnectorSpec)).when(configPersistence).getConfig(eq(ConfigSchema.STANDARD_SOURCE_DEFINITION), any(), any());
+
+    when(featureFlags.exposeSecretsInExport()).thenReturn(false);
+    writeSourceWithSourceConnection(configPersistence, SOURCE_GITHUB);
+    writeSourceWithSourceConnection(configPersistence, SOURCE_POSTGRES);
+    writeDestinationWithDestinationConnection(configPersistence, DESTINATION_S3);
+    final Map<String, Stream<JsonNode>> result = configPersistence.dumpConfigs();
```
[the remark this answers is not in the document]
Comment (benmoriceau) [reply airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/DatabaseConfigPersistenceTest.java:195]: The mocking will be a big pain, I'll need to create a JsonNode for the config spec that will match make sure that the secret will be sanitized and also ensure that the connection definition have a secret to sanitize.
This is tested in the test of the JsonSecretsProcessor.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java (hunk 6 above)
Comment (benmoriceau) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java]: Done

**Comment by cgardens** [comment]
cool! lgtm

**Review by michel-tricot — COMMENTED** [review]
The implementation using the boolean to branch in the code creates a different behavior with how we process the config. if for any reason in the future we apply some additional, non-secret related transformation on the config, we will have two places to modify or we might just forget...

I suggest instead that we don't use a feature flag in the configPersistence class but instead we instantiate a different type of JsonSecretProcessor that is a pass through.

**Comment by benmoriceau** [comment]
> The implementation using the boolean to branch in the code creates a different behavior with how we process the config. if for any reason in the future we apply some additional, non-secret related transformation on the config, we will have two places to modify or we might just forget...
> 
> I suggest instead that we don't use a feature flag in the configPersistence class but instead we instantiate a different type of JsonSecretProcessor that is a pass through.

I am not sure how having another class will help here. In our use case, it will be a NoOpJsonProcess that will just return the same JsonNode.
I agree that we might have to modify 2 places in the future, here is what I'm gonna do:
- set `maskSecrets` as private (or default for testing)
- create a `transform` method in the `JsonSecretProcessor`. This method will take booleans to know which transformation are needed. For now it is only the secret masking. If future method appears, it will be called through the transform method and won't be duplicated.

Does it sound like a good approach @michel-tricot ?

**Comment by benmoriceau** [comment]
> > The implementation using the boolean to branch in the code creates a different behavior with how we process the config. if for any reason in the future we apply some additional, non-secret related transformation on the config, we will have two places to modify or we might just forget...
> > I suggest instead that we don't use a feature flag in the configPersistence class but instead we instantiate a different type of JsonSecretProcessor that is a pass through.
> 
> I am not sure how having another class will help here. In our use case, it will be a NoOpJsonProcess that will just return the same JsonNode. I agree that we might have to modify 2 places in the future, here is what I'm gonna do:
> 
> * set `maskSecrets` as private (or default for testing)
> * create a `transform` method in the `JsonSecretProcessor`. This method will take booleans to know which transformation are needed. For now it is only the secret masking. If future method appears, it will be called through the transform method and won't be duplicated.
> 
> Does it sound like a good approach @michel-tricot ?

Will do it in another PR in order to avoid over loading that one.

**Pull request merged**